pkg/controller/elasticsearch/driver/upgrade_predicates.go

```diff
@@ -218,6 +222,15 @@ var predicates = [...]Predicate{
 			deletedPods []corev1.Pod,
 			maxUnavailableReached bool,
 		) (b bool, e error) {
+			health, err := context.esState.Health()
+			if err != nil {
+				return false, err
+			}
+			_, healthyNode := context.healthyPods[candidate.Name]
+			if health.NumberOfNodes == 1 && health.Status == esv1.ElasticsearchYellowHealth && healthyNode {
```

Comment:
What I meant was that we are basing the decision on the result of `_cat/health` or `_cluster/health` here. But the internal view of cluster membership might not be consistent with the external state of things in Kubernetes. The operator will take care to do node deletions and creations before the rolling upgrade. But Pods in a multi-node cluster might be not `Running` for other reasons in which case we would proceed with the upgrade assuming it is a single node cluster and potentially drop into `red` health. I think we need to look at the actual Pods in the Kubernetes cluster e.g. with `StatefulSetList.GetActualPods`

Reply:
understood, while reviewing the code I thought that the best option would be to break a bit the Open-Closed Principle. 

Checking `predicateContext` we have nothing that could allow us to know the total number of pods on the `StatefulSet`, so I decided to add new information there (sorry OCP). 

We can add `GetActualPods` on `upgrade_pods_deletion.go` or in `upgrade.go>handleRollingUpgrades`, upgrade.go seems a better option as in the function `handleRollingUpgrades` we already calculate information as healthyPods, actualMasters.. etc.

I know I'm taking a risk and I can change it if needed.